In [97]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [98]:
# Load the given datasets
df_cat = pd.read_csv('chronic_kidney_disease_categorical.csv')
df_numerical = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
print(df_cat.head())
print('------------------------------------------------')
print(df_numerical.head())

   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  
------------------------------------------------
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

## Answer: 
Features are characteristics of a patient collected by a doctor, such as the patient's age, red blood cell health, whether they test positive for conditions like diabetes or coronary artery disease, and their appetite. The labels represent whether the patient has chronic kidney disease or not.

Across both datasets, there are both categorical and numerical features. Examples of categorical features include appetite (appet), red blood cells (rbc), and pus cells (pc). Examples of numerical features include age, blood pressure (bp), and blood glucose random (bgr).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [99]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates = df_numerical.duplicated(subset='unique_id', keep=False)
duplicate_count = duplicates.sum()
if duplicate_count > 0:
    print(f"{duplicate_count} duplicates in numerical dataset")

# Drop these duplicate rows
df_numerical = df_numerical.drop_duplicates(subset='unique_id', keep='first')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates = df_cat.duplicated(subset='unique_id', keep=False)
duplicate_count = duplicates.sum()
if duplicate_count > 0:
    print(f"{duplicate_count} duplicates in categorical dataset")
print('------------------------------------------------')
print("duplicate rows: ")
print(f"{df_cat[duplicates]}")
print('------------------------------------------------')
df_cat = df_cat.drop_duplicates(subset='unique_id', keep='first')

59 duplicates in numerical dataset
59 duplicates in categorical dataset
------------------------------------------------
duplicate rows: 
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
10      708255  2.0  0.0    normal    normal  notpresent  notpresent   no   
34      441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
39      240975  0.0  0.0    normal    normal  notpresent  notpresent   no   
51      354223  0.0  0.0    normal    normal  notpresent  notpresent   no   
71      283427  0.0  0.0    normal    normal  notpresent  notpresent   no   
74      302906  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
88      856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
94      501321  1.0  0.0       NaN    normal  notpresent  notpresent   no   
113     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
124     261137 

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all unique ids are present in both datasets. Based on output from the cell below, there are 400 unique ids in the categorical data, and 390 unique ids in the numerical data. Rows that are missing from one of the datasets have NaN values in the columns belonging to the datasets which the row's id is missing from. 

In [100]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_df = pd.merge(df_cat, df_numerical, on='unique_id', how='outer')

#Print the combined dataset
print(merged_df.head())

# print number of unique ids in both datasets
categorical_ids = df_cat['unique_id'].tolist()
numerical_ids = df_numerical['unique_id'].tolist()
print(f'{len(categorical_ids)} unique ids in categorical data, {len(numerical_ids)} unique ids in numerical data')

# find id in categorical_ids not in numerical ids, print it
categorical_only_ids = list(set(categorical_ids) - set(numerical_ids))
print("\nrow of merged dataframe of id only in categorical data:")
print(merged_df[merged_df['unique_id'] == categorical_only_ids[0]])

   unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0     102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1     103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2     104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3     105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4     107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target   age    bp    bgr     bu    sc    sod  pot  \
0  no  poor  yes  yes     ckd  59.0  70.0   76.0  186.0  15.0  135.0  7.6   
1  no  good   no   no  notckd  24.0  80.0  125.0    NaN   NaN  136.0  3.5   
2  no  good   no   no     ckd  65.0  80.0  215.0  133.0   2.5    NaN  NaN   
3  no  good   no   no     ckd  90.0  90.0  139.0   89.0   3.0  140.0  4.1   
4  no  good   no   no  notckd  60.0  70.0    NaN    NaN   NaN    NaN  NaN   

   hemo   pcv     wbcc  rbcc  
0   7.1  22.0   3800.0   2.1  
1  15.

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [101]:
# Calculate the percentage of rows that contain atleast one missing value
num_rows_with_nan = merged_df.isna().any(axis=1).sum()

# Print %
percent_nans = 100 * num_rows_with_nan / len(merged_df)
print(percent_nans)

# Drop these rows from the dataset
merged_df = merged_df.dropna().reset_index(drop=True)

# Print the Dataset
print(merged_df)

61.75
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       109053  0.0  0.0    normal    normal  notpresent  notpresent   no   
2       114717  0.0  0.0    normal    normal  notpresent  notpresent   no   
3       118191  0.0  0.0    normal    normal  notpresent  notpresent   no   
4       118596  0.0  0.0    normal    normal  notpresent  notpresent   no   
5       124304  3.0  4.0    normal  abnormal  notpresent  notpresent  yes   
6       125587  0.0  0.0    normal    normal  notpresent  notpresent   no   
7       129053  4.0  0.0  abnormal  abnormal  notpresent     present   no   
8       129654  0.0  0.0    normal    normal  notpresent  notpresent   no   
9       131563  0.0  0.0    normal    normal  notpresent  notpresent   no   
10      137148  3.0  2.0  abnormal  abnormal     present  notpresent  yes   
11      153227  3.0  5.0  abnormal    normal  notpresent     present  

## Part 3.4: Sort the dataset according to the Labels

In [102]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = merged_df.sort_values(by='Target').reset_index(drop=True)

# Print the dataset
print(sorted_df)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
4       353025  3.0  4.0    normal    normal  notpresent  notpresent  yes   
5       397388  2.0  2.0    normal    normal  notpresent     present  yes   
6       428486  0.0  0.0    normal    normal  notpresent  notpresent  yes   
7       438182  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
8       474407  4.0  0.0    normal  abnormal     present     present   no   
9       475200  4.0  2.0  abnormal  abnormal  notpresent     present  yes   
10      481293  3.0  1.0    normal  abnormal     present  notpresent  yes   
11      484175  2.0  0.0  abnormal  abnormal     present  notpresent  yes   

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [103]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
dummies = pd.get_dummies(sorted_df[categorical_columns], dtype='int')
df_with_cat_dummies = pd.concat([sorted_df, dummies], axis=1).drop(columns=categorical_columns)

# Print the dataset
print(df_with_cat_dummies)

     unique_id   al   su   age     bp    bgr     bu    sc    sod   pot  hemo  \
0       102677  3.0  0.0  59.0   70.0   76.0  186.0  15.0  135.0   7.6   7.1   
1       337834  4.0  0.0  48.0   70.0  117.0   56.0   3.8  111.0   2.5  11.2   
2       343710  2.0  0.0  61.0   80.0  173.0  148.0   3.9  135.0   5.2   7.7   
3       349892  1.0  3.0  59.0   70.0  424.0   55.0   1.7  138.0   4.5  12.6   
4       353025  3.0  4.0  40.0   70.0  253.0  150.0  11.9  132.0   5.6  10.9   
5       397388  2.0  2.0  63.0  100.0  280.0   35.0   3.2  143.0   3.5  13.0   
6       428486  0.0  0.0  73.0   70.0   70.0   32.0   0.9  125.0   4.0  10.0   
7       438182  2.0  0.0  73.0   80.0  253.0  142.0   4.6  138.0   5.8  10.5   
8       474407  4.0  0.0  21.0   90.0  107.0   40.0   1.7  125.0   3.5   8.3   
9       475200  4.0  2.0  64.0  100.0  163.0   54.0   7.2  140.0   4.6   7.9   
10      481293  3.0  1.0  60.0   60.0  288.0   36.0   1.7  130.0   3.0   7.9   
11      484175  2.0  0.0  53.0   90.0   

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

In the downstream modelling, modelling a binary variable with jut one column would be equivalent, since the variable can only take on two values, so if the column is a false for a datapoint with some features, then the other value must be satisfied. If the categorical variable took on more than 2 values, modelling the variable with just one column would not work. Assuming that the set of all five values is known, four binary features could be used to model the 5-value categorical feature. However, in the case of condiment types, we do not necessarily know all 5 different types of condiments, and therefore this approach would, to a large extent, not be applicable.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [104]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
outliers_removed_df = df_with_cat_dummies
for col in numerical_columns:
    mean = df_with_cat_dummies[col].mean()
    std_dev = df_with_cat_dummies[col].std()
    upper_bound = mean + 3 * std_dev
    lower_bound = mean - 3 * std_dev
    outliers_removed_df = outliers_removed_df[(outliers_removed_df[col] >= lower_bound) & (outliers_removed_df[col] <= upper_bound)]
outliers_removed_df = outliers_removed_df.reset_index(drop=True)

# Print the dataset
print(outliers_removed_df)

     unique_id   al   su   age     bp    bgr     bu   sc    sod  pot  hemo  \
0       343710  2.0  0.0  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7   
1       397388  2.0  2.0  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0   
2       438182  2.0  0.0  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5   
3       474407  4.0  0.0  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3   
4       475200  4.0  2.0  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9   
5       481293  3.0  1.0  60.0   60.0  288.0   36.0  1.7  130.0  3.0   7.9   
6       514721  2.0  0.0  56.0   90.0  129.0  107.0  6.7  131.0  4.8   9.1   
7       995177  4.0  3.0  69.0   70.0  214.0   96.0  6.3  120.0  3.9   9.4   
8       546225  1.0  0.0  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8   
9       980291  3.0  0.0  62.0   70.0  122.0   42.0  1.7  136.0  4.7  12.6   
10      955830  3.0  1.0  55.0   80.0  214.0   73.0  3.9  137.0  4.9  10.9   
11      923613  4.0  1.0  64.0   60.0  239.0   58.0  4.3  137.0 

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [105]:
# Normalize the all Numerical Attributes in the dataset.
normalized_df = outliers_removed_df
normalized_df[numerical_columns] = (normalized_df[numerical_columns] - normalized_df[numerical_columns].min()) / \
                                    (normalized_df[numerical_columns].max() - normalized_df[numerical_columns].min())

# Print the dataset
print(normalized_df)

     unique_id   al   su       age    bp       bgr        bu        sc  \
0       343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
1       397388  2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679   
2       438182  2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519   
3       474407  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494   
4       475200  4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506   
5       481293  3.0  1.0  0.729730  0.00  0.935622  0.169935  0.160494   
6       514721  2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778   
7       995177  4.0  3.0  0.851351  0.25  0.618026  0.562092  0.728395   
8       546225  1.0  0.0  0.540541  0.00  0.399142  0.535948  0.358025   
9       980291  3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494   
10      955830  3.0  1.0  0.662162  0.50  0.618026  0.411765  0.432099   
11      923613  4.0  1.0  0.783784  0.00  0.725322  0.313725  0.481481   
12      843362  4.0  0.0  0.878378  0.

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them


Answer:
Yes. unique_id is not appropriate for modeling. The unique id is assigned via some mechanism to each patient, and its relationship to whether somebody has kidney disease or not is not important, as the goal of modeling this dataset would be to predict whether somebody has kidney disease based on observable biological characteristics of that person. Furthermore, the unique_id column is identifying personal information, and could be a privacy risk if used in a publicly available model or used in publicly available statistics.

In [106]:
#Remove that column
unnecessary_cols_removed_df = normalized_df.drop(columns=['unique_id'])

# Print the dataset
print(unnecessary_cols_removed_df)

      al   su       age    bp       bgr        bu        sc       sod  \
0    2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1    2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2    2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3    4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4    4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
5    3.0  1.0  0.729730  0.00  0.935622  0.169935  0.160494  0.333333   
6    2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
7    4.0  3.0  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
8    1.0  0.0  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
9    3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
10   3.0  1.0  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
11   4.0  1.0  0.783784  0.00  0.725322  0.313725  0.481481  0.566667   
12   4.0  0.0  0.878378  0.00  0.206009  0.751634  

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [107]:
# Export the dataframe to a new csv file
cleaned_df = unnecessary_cols_removed_df
cleaned_df.to_csv('cleaned_kidney_data.csv')

# Export the dataframe to a new json file
cleaned_df.to_json('cleaned_kidney_data.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here
|  al  |  su  |       age |    bp |    bgr |        bu |        sc |      sod |      pot |      hemo |       pcv |      wbcc |      rbcc |  rbc_abnormal |  rbc_normal |  pc_abnormal |  pc_normal |  pcc_notpresent |  pcc_present |  ba_notpresent |  ba_present |  htn_no |  htn_yes |  dm_no |  dm_yes |  cad_no |  cad_yes |  appet_good |  appet_poor |  pe_no |  pe_yes |  ane_no |  ane_yes |  Target_ckd |  Target_notckd |
|------|------|-----------|-------|--------|-----------|-----------|----------|----------|------------|-----------|-----------|-----------|----------------|-------------|---------------|------------|----------------|--------------|----------------|-------------|---------|----------|--------|---------|---------|----------|-------------|-------------|--------|---------|---------|----------|--------------|----------------|
|  2.0 |  0.0 | 0.743243  |  0.50 | 0.442060 | 0.901961 | 0.432099 | 0.500000 | 0.657143 | 0.172131 | 0.210526 | 0.395161 | 0.153846 |              1 |           0 |             1 |          0 |                1 |            0 |              1 |           0 |       0 |        1 |      0 |       1 |       0 |        1 |           0 |           1 |      0 |       1 |       0 |        1 |            1 |              0 |
|  2.0 |  2.0 | 0.770270  |  1.00 | 0.901288 | 0.163399 | 0.345679 | 0.766667 | 0.171429 | 0.606557 | 0.631579 | 0.443548 | 0.410256 |              0 |           1 |             0 |          1 |                1 |            0 |              0 |           1 |       0 |        1 |      1 |       0 |       0 |        1 |           1 |           0 |      1 |       0 |       0 |        1 |            1 |              0 |
|  2.0 |  0.0 | 0.905405  |  0.50 | 0.785408 | 0.862745 | 0.518519 | 0.600000 | 0.828571 | 0.401639 | 0.447368 | 0.233871 | 0.435897 |              1 |           0 |             1 |          0 |                1 |            0 |              1 |           0 |       0 |        1 |      0 |       1 |       0 |        1 |           1 |           0 |      0 |       1 |       1 |        0 |            1 |              0 |
|  4.0 |  0.0 | 0.202703  |  0.75 | 0.158798 | 0.196078 | 0.160494 | 0.166667 | 0.171429 | 0.221311 | 0.184211 | 0.653226 | 0.333333 |              0 |           1 |             1 |          0 |                0 |            1 |              0 |           1 |       1 |        0 |      1 |       0 |       1 |        0 |           1 |           0 |      0 |       1 |       0 |        1 |            1 |              0 |
|  4.0 |  2.0 | 0.783784  |  1.00 | 0.399142 | 0.287582 | 0.839506 | 0.666667 | 0.485714 | 0.188525 | 0.263158 | 0.258065 | 0.205128 |              1 |           0 |             1 |          0 |                1 |            0 |              1 |           1 |       0 |        1 |      0 |       1 |       1 |        0 |           1 |           0 |      1 |       0 |       0 |        1 |            1 |              0 |
|  3.0 |  1.0 | 0.729730  |  0.00 | 0.935622 | 0.169935 | 0.160494 | 0.333333 | 0.028571 | 0.188525 | 0.236842 | 0.879032 | 0.102564 |              0 |           1 |             1 |          0 |                0 |            1 |              1 |           0 |       0 |        1 |      1 |       0 |       0 |        1 |           0 |           0 |      1 |       0 |       0 |        1 |            1 |              0 |
|  2.0 |  0.0 | 0.675676  |  0.75 | 0.253219 | 0.633987 | 0.777778 | 0.366667 | 0.542857 | 0.286885 | 0.342105 | 0.169355 | 0.205128 |              1 |           0 |             1 |          0 |                1 |            0 |              1 |           0 |       0 |        1 |      1 |       0 |       0 |        1 |           0 |           0 |      1 |       0 |       0 |        1 |            1 |              0 |
|  4.0 |  3.0 | 0.851351  |  0.25 | 0.618026 | 0.562092 | 0.728395 | 0.000000 | 0.285714 | 0.311475 | 0.315789 | 0.580645 | 0.179487 |              0 |           1 |             1 |          0 |                0 |            1 |              1 |           0 |       0 |        1 |      1 |       0 |       0 |        1 |           0 |           0 |      0 |       1 |       0 |        1 |            1 |              0 |
|  1.0 |  0.0 | 0.540541  |  0.00 | 0.399142 | 0.535948 | 0.358025 | 0.700000 | 0.314286 | 0.344262 | 0.315789 | 0.830645 | 0.153846 |              0 |           1 |             0 |          1 |                1 |            0 |              1 |           0 |       0 |        1 |      0 |       1 |       1 |        0 |           1 |           0 |      0 |       1 |       0 |        1 |            1 |              0 |
|  3.0 |  0.0 | 0.756757  |  0.25 | 0.223176 | 0.209150 | 0.160494 | 0.533333 | 0.514286 | 0.573770 | 0.605263 | 0.290323 | 0.333333 |              0 |           1 |             1 |          0 |                1 |            0 |              1 |           0 |       0 |        1 |      0 |       1 |       1 |        0 |           1 |           0 |      0 |       1 |       0 |        1 |            1 |              0 |
|  3.0 |  1.0 | 0.662162  |  0.50 | 0.618026 | 0.411765 | 0.432099 | 0.566667 | 0.571429 | 0.434426 | 0.473684 | 0.250000 | 0.282051 |              0 |           1 |             1 |          0 |                1 |            0 |              1 |           0 |       0 |        1 |      0 |       1 |       1 |        0 |           1 |           0 |      0 |       1 |       0 |        1 |            1 |              0 |
|  4.0 |  1.0 | 0.783784  |  0.00 | 0.725322 | 0.313725 | 0.481481 | 0.566667 | 0.714286 | 0.319672 | 0.342105 | 0.258065 | 0.205128 |              1 |           0 |             1 |          0 |                1 |            0 |              0 |           1 |       0 |        1 |      0 |       1 |       1 |        0 |           0 |           1 |      0 |       1 |       0 |        1 |            1 |              0 |
|  4.0 |  0.0 | 0.878378  |  0.00 | 0.206009 | 0.751634 | 0.604938 | 0.533333 | 0.571429 | 0.475410 | 0.500000 | 0.879032 | 0.435897 |              0 |           1 |             0 |          1 |                1 |            0 |              1 |           0 |       0 |        1 |      0 |       1 |       1 |        0 |           0 |           1 |      0 |       1 |       0 |        1 |            1 |              0 |
|  3.0 |  0.0 | 0.878378  |  0.25 | 0.639485 | 0.470588 | 0.395062 | 0.433333 | 0.428571 | 0.393443 | 0.447368 | 0.104839 | 0.256410 |              0 |           1 |             1 |          0 |                0 |            1 |              0 |           1 |       0 |        1 |      1 |       0 |       0 |        1 |           0 |           0 |      1 |       0 |       0 |        1 |            1 |              0 |
|  1.0 |  0.0 | 0.716216  |  0.50 | 1.000000 | 0.163399 | 0.111111 | 0.066667 | 0.171429 | 0.393443 | 0.500000 | 0.532258 | 0.435897 |              1 |           0 |             0 |          1 |                1 |            0 |              1 |           0 |       0 |        1 |      0 |       1 |       1 |        0 |           1 |           0 |      1 |       0 |       0 |        1 |            1 |              0 |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

If an llm detects that a user is asking it to do a data conversion, it could use a parser to check it's own output before sending it back to the user. Another piece of software partially relying on LLMs could be designed to check that the output actually matches the user's prompt.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [108]:
# Code to rename all the columns in the dataset
full_column_names = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_abnormal': 'Red Blood Cells (Abnormal)',
    'rbc_normal': 'Red Blood Cells (Normal)',
    'pc_abnormal': 'Pus Cell (Abnormal)',
    'pc_normal': 'Pus Cell (Normal)',
    'pcc_present': 'Pus Cell Clumps (Present)',
    'pcc_notpresent': 'Pus Cell Clumps (Not Present)',
    'ba_present': 'Bacteria (Present)',
    'ba_notpresent': 'Bacteria (Not Present)',
    'htn_yes': 'Hypertension (Yes)',
    'htn_no': 'Hypertension (No)',
    'dm_yes': 'Diabetes Mellitus (Yes)',
    'dm_no': 'Diabetes Mellitus (No)',
    'cad_yes': 'Coronary Artery Disease (Yes)',
    'cad_no': 'Coronary Artery Disease (No)',
    'appet_good': 'Appetite (Good)',
    'appet_poor': 'Appetite (Poor)',
    'pe_yes': 'Pedal Edema (Yes)',
    'pe_no': 'Pedal Edema (No)',
    'ane_yes': 'Anemia (Yes)',
    'ane_no': 'Anemia (No)',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'Target_ckd': 'Chronic Kidney Disease (yes)',
    'Target_notckd': 'Chronic Kidney Disease (no)'
}

# Rename the columns in the dummies DataFrame
cleaned_df = cleaned_df.rename(columns=full_column_names)

# Display the new DataFrame with full names
print(cleaned_df)

     Albumin  Sugar       Age  Blood Pressure  Blood Glucose Random  \
0        2.0    0.0  0.743243            0.50              0.442060   
1        2.0    2.0  0.770270            1.00              0.901288   
2        2.0    0.0  0.905405            0.50              0.785408   
3        4.0    0.0  0.202703            0.75              0.158798   
4        4.0    2.0  0.783784            1.00              0.399142   
5        3.0    1.0  0.729730            0.00              0.935622   
6        2.0    0.0  0.675676            0.75              0.253219   
7        4.0    3.0  0.851351            0.25              0.618026   
8        1.0    0.0  0.540541            0.00              0.399142   
9        3.0    0.0  0.756757            0.25              0.223176   
10       3.0    1.0  0.662162            0.50              0.618026   
11       4.0    1.0  0.783784            0.00              0.725322   
12       4.0    0.0  0.878378            0.00              0.206009   
13    

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [109]:
df_with_target_col = cleaned_df.rename(columns={'Chronic Kidney Disease (yes)': 'Target'}) # rename column of cleaned df to be 'Target'

# Converted SQL to Pandas code
output_of_sql_query = df_with_target_col.groupby('Target').size().reset_index(name='count')
print(output_of_sql_query)

   Target  count
0       0    111
1       1     23


*Describe what the above code does here*

This code creates a new table, with a row for each unique categorical value in the `Target` column of the table `your_dataframe`. The unique value of the row is stored in a column called `Target`, and the number of rows in the dataframe `your_dataframe` which have that value in their `Target` column is stored in the `count` column for that row.